File: app/src/main/java/com/mcal/disassembler/view/FloatingMenu.java
package com.mcal.disassembler.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.PixelFormat;
import android.os.Build;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;
import android.view.WindowManager;

import com.mcal.disassembler.util.Utils;

public class FloatingMenu {
    public static int xPos = 0;
    public static int yPos = 0;
    private final Context context;
    private final String path;
    public boolean isAdded = false;
    public WindowManager wm;
    public WindowManager.LayoutParams params;
    public FloatingMenuView floatView;

    public FloatingMenu(Context c, String filePath) {
        context = c;
        path = filePath;
    }

    @SuppressLint("ClickableViewAccessibility")
    public void show() {
        int LAYOUT_FLAG;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            LAYOUT_FLAG = WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY;
        } else {
            LAYOUT_FLAG = WindowManager.LayoutParams.TYPE_PHONE;
        }

        wm = (WindowManager) context.getApplicationContext().getSystemService(Context.WINDOW_SERVICE);
        params = new WindowManager.LayoutParams();

        params.type = LAYOUT_FLAG; // WindowManager.LayoutParams.TYPE_SYSTEM_ALERT;

        params.format = PixelFormat.RGBA_8888;

        params.flags = WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL | WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE;
        params.height = wm.getDefaultDisplay().getHeight() / 2;
        params.width = Utils.dp(context, 250);

        params.x = xPos;
        params.y = yPos;

        floatView = new FloatingMenuView(context, this, path, params.width, params.height);
        floatView.setClickable(true);

        floatView.setOnTouchListener(new OnTouchListener() {
            int lastX, lastY;
            int paramX, paramY;

            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        lastX = (int) event.getRawX();
                        lastY = (int) event.getRawY();
                        paramX = params.x;
                        paramY = params.y;
                        break;
                    case MotionEvent.ACTION_MOVE:
                        int dx = (int) event.getRawX() - lastX;
                        int dy = (int) event.getRawY() - lastY;
                        params.x = paramX + dx;
                        params.y = paramY + dy;
                        wm.updateViewLayout(floatView, params);
                        break;
                }
                FloatingMenu.xPos = params.x;
                FloatingMenu.yPos = params.y;
                return false;
            }
        });

        wm.addView(floatView, params);
        isAdded = true;
    }

    public void dismiss() {
        wm.removeView(floatView);
    }
}

File: app/src/main/java/com/mcal/disassembler/view/FloatingMenu.kt
package com.mcal.disassembler.view

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.PixelFormat
import android.os.Build
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.view.WindowManager
import com.mcal.disassembler.App

class FloatingMenu(
    private val activity: Activity,
    path: String
) {
    private var isAdded = false
    private val windowManager =
        activity.applicationContext.getSystemService(Context.WINDOW_SERVICE) as WindowManager
    private val params = WindowManager.LayoutParams()
    private val floatView = FloatingMenuView(activity, this, path)

    fun show() {
        windowManager.addView(floatView.apply {
            isClickable = true
            setOnTouchListener(object : OnTouchListener {
                var lastX = 0
                var lastY = 0
                var paramX = 0
                var paramY = 0

                @SuppressLint("ClickableViewAccessibility")
                override fun onTouch(v: View, event: MotionEvent): Boolean {
                    val param = params
                    when (event.action) {
                        MotionEvent.ACTION_DOWN -> {
                            lastX = event.rawX.toInt()
                            lastY = event.rawY.toInt()
                            paramX = param.x
                            paramY = param.y
                        }

                        MotionEvent.ACTION_MOVE -> {
                            val dx = event.rawX.toInt() - lastX
                            val dy = event.rawY.toInt() - lastY
                            param.x = paramX + dx
                            param.y = paramY + dy
                            windowManager.updateViewLayout(this@apply, param)
                        }
                    }
                    xPos = param.x
                    yPos = param.y
                    return false
                }
            })
        }, params.apply {
            type = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
            } else {
                WindowManager.LayoutParams.TYPE_PHONE
            }
            format = PixelFormat.TRANSPARENT
            flags = WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
            width = App.dp(activity, 250)
            height = windowManager.defaultDisplay.height / 2
            x = xPos
            y = yPos
        })
        isAdded = true
    }

    fun dismiss() {
        windowManager.removeView(floatView)
    }

    companion object {
        private var xPos = 0
        private var yPos = 0
    }
}
